ouroboros-consensus/src/Ouroboros/Consensus/Util/MonadSTM/StrictMVar.hs

```diff
@@ -0,0 +1,151 @@
+{-# LANGUAGE BangPatterns   #-}
+{-# LANGUAGE NamedFieldPuns #-}
+
+module Ouroboros.Consensus.Util.MonadSTM.StrictMVar (
+    StrictMVar
+  , newMVar
+  , newMVarWithInvariant
+  , newEmptyMVar
+  , newEmptyMVarWithInvariant
+  , takeMVar
+  , tryTakeMVar
+  , putMVar
+  , tryPutMVar
+  , readMVar
+  , readMVarSTM
+  , tryReadMVar
+  , swapMVar
+  , isEmptyMVar
+  , updateMVar
+  ) where
+
+import           Control.Monad (when)
+import           Control.Monad.Class.MonadSTM (MonadSTM (..))
+import qualified Control.Monad.Class.MonadSTM as Lazy
+import           Control.Monad.Class.MonadSTM.Strict (checkInvariant)
+import           GHC.Stack
+
+{-------------------------------------------------------------------------------
+  Strict MVar
+-------------------------------------------------------------------------------}
+
+-- | Strict MVar (modelled using a lazy 'Lazy.TMVar' under the hood)
+--
+-- The 'StrictMVar' API is slightly stronger than the usual 'MVar' one, as we
+-- offer a primitive to read the value of the MVar even if it is empty (in which
+-- case we will return the oldest known stale one). See 'readMVarSTM'.
+--
+-- There is a weaker invariant for a 'StrictMVar' than for a 'StrictTVar':
+-- although all functions that modify the 'StrictMVar' check the invariant, we
+-- do /not/ guarantee that the value inside the 'StrictMVar' always satisfies
+-- the invariant. Instead, we /do/ guarantee that if the 'StrictMVar' is updated
+-- with a value that does not satisfy the invariant, an exception is thrown. The
+-- reason for this weaker guarantee is that leaving an 'MVar' empty can lead to
+-- very hard to debug "blocked indefinitely" problems.
+--
+-- This is also the reason we do not offer support for an invariant in
+-- 'StrictTMVar': if we throw an exception from an STM transaction, the STM
+-- transaction is not executed, and so we would not even be able to provide the
+-- weaker guarantee that we provide for 'StrictMVar'.
+data StrictMVar m a = StrictMVar
+  { invariant :: !(a -> Maybe String)
+    -- ^ Invariant checked whenever updating the 'StrictMVar'.
+  , tmvar     :: !(Lazy.LazyTMVar m a)
+    -- ^ The main TMVar supporting this 'StrictMVar'
+  , tvar      :: !(Lazy.LazyTVar m a)
+    -- ^ TVar for supporting 'readOldValue'
+    --
+    -- This TVar is always kept up to date with the 'LazyTMVar', but holds on
+    -- the old value of the 'LazyTMVar' when it is empty. This is very useful
+    -- to support single writer/many reader scenarios.
+    --
+    -- NOTE: We should always update the 'tmvar' before the 'tvar' so that if
```

Comment:
Hmm, why lower-case + `''`? This will give Haddock errors. Either use `TMVar`/`TVar` or `@` (or no delimiters).